### PR TITLE
Add support for PrimeKey Hardware Appliance

### PR DESCRIPTION
**Pull request by bl3nd3r**
This pull request is to add support for PrimeKey hardware appliances.

- [EJBCA Appliance](https://doc.primekey.com/ejbca-appliance)
- [SignServer Appliance](https://doc.primekey.com/signserver-appliance)

I understand both appliances use the same MiBs, described [here](https://doc.primekey.com/ejbca-appliance/operations/webconf-configurator-of-hardware-appliance/monitoring).

As per the above link a single bulkwalk will not reach all OIDs on these appliances.
> All OIDs can be reached by the following snmpwalk commands. Replace the IP address in each command with the one of your system:
> 
> ```
> # for the standard group
> snmpwalk -v2c -On -c public 192.168.5.162
> # for the system group
> snmpwalk -v2c -On -c public 192.168.5.162 .1.3.6.1.4.1.22408.1.1.2.1
> # for the HSM group
> snmpwalk -v2c -On -c public 192.168.5.162 .1.3.6.1.4.1.22408.1.1.2.2
> ```
> 

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: includes/discovery/sensors/load/primekey.inc.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        'oid'   => '.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.52.1',
+        'group' => 'CPU',
+    ],
+];
+
+$class = 'load';
+
+$type = 'primekey';
+$divisor = 1;
+$multiplier = 100;
```
Comment (bl3nd3r) [inline includes/discovery/sensors/load/primekey.inc.php:43]: Here we're checking the system load averages as reported by "uptime" in Linux, generally a number between 0 and 1.
We multiply by 100 which should provide approximately a percentage - assuming a single threaded system.

The appliance I'm testing against apparently has an "Intel(R) Xeon(R) CPU E3-1225 v5" processor - i.e. 4 Cores, 4 threads.

On that bases I'm not sure if I should set the multiplier to 25, on the basis that a system load of 1.0 would indicate that the CPU is 25% (one quarter) full.

Comment (PipoCanaja) [reply includes/discovery/sensors/load/primekey.inc.php:43]: Hi
System load is a value "per se" and should not be modified.

Comment (bl3nd3r) [reply includes/discovery/sensors/load/primekey.inc.php:43]: Load is displayed by Librenms as a percentage, with certain default warning and thresholds

Comment (PipoCanaja) [reply includes/discovery/sensors/load/primekey.inc.php]: Load as sensor in LibreNMS is a percentage and is used for UPSes etc.

Usually, CPU usage is reported with the "processor" module and not the "sensor" module. In the processor module, you can choose the source of the data, and the ratio to apply on it. 

So, 2 variants are: 
- Either you use the load as reported by the linux device, and display it with a "count" sensor, without changing it. 
- Or you use the processor module and use the load value with a maximum defined as "cpucount". BUT if you do so, then any load above cpucount will result in a percentage over 100% which does not make sense. 

That's why a load value should be kept as it is.

Comment (bl3nd3r) [reply includes/discovery/sensors/load/primekey.inc.php]: Checked with the vendor and for now I think setting "cpucount" to 4 will work for all the appliances they sell currently. 

I've not yet found another reference to cpucount in this repo, however on [this page](https://docs.librenms.org/Developing/os/Mem-CPU-Information/#yaml_1) I see a yaml data key called precision:

> <html>
<body>


precision | 1 | The multiplier to multiply the data by. If this is negative, the data will be multiplied then subtracted from 100.
-- | -- | --



</body>
</html>
Having taken a quick rummage around the LibreNMS/OS folder I think I need to find and better comprehend the Processor::discover function before I can go down the processor module route.....

TL;DR: Should I pick your first option and display it as an unaltered "count" sensor?  😁

**Comment by bl3nd3r** [comment]
CPU load statistics moved from Load to Count.

Alert thresholds not set to avoid later confusion.

**Comment by PipoCanaja** [comment]
Hi @bl3nd3r 
Could you try to use as much YAML file as possible instead of PHP files ? When possible, YAML files are easier to maintain in the longterm than PHP files. 
Thanx

**Comment by bl3nd3r** [comment]
Hi @PipoCanaja 

I agree YAML would be easier to maintain. My attempts to convert from PHP to YAML have to date been unsuccessful despite having read the documentation and making a few attempts.

I think I'm struggling here with a possibly a combination of one or more of the following:
1. Misinterpreting the examples in the docs and/or other yaml files I've looked at.
2. Not knowing where in the php code the YAML files get ingested so I can add some debug statements to figure out where I've gone wrong.
3. Possibly the MiB files don't contain what the docs refer to as a "sane table that can be traversed and contains all of the data".

In any case if you or someone else can demonstrate how these particular MiB files can be read via YAML I'd be more than happy to learn by example.

**Comment by PipoCanaja** [comment]
@bl3nd3r 
You are right, it seems that this SNMP implementation and the MIB file that goes with it are badly structured. But it is most probably possible to do something anyway. It is difficult to try (and fail) without having the device in front of my keyboard but I'll try to come with something later.

**Comment by bl3nd3r** [comment]
In case it helps at all here's what a walk looks like.

```
[librenms@librenms ~]$snmpwalk -v2c -On -c $COMMUNITY $HOST
.1.3.6.1.2.1.1.1.0 = STRING: Linux vgw 4.19.46 #1 SMP Thu Sep 19 02:32:09 CEST 2019 x86_64
.1.3.6.1.2.1.1.2.0 = OID: .1.3.6.1.4.1.8072.3.2.10
.1.3.6.1.2.1.1.3.0 = Timeticks: (67701531) 7 days, 20:03:35.31
.1.3.6.1.2.1.1.4.0 = STRING: PrimeKey <support@primekey.com>
.1.3.6.1.2.1.1.5.0 = STRING: vgw
.1.3.6.1.2.1.1.6.0 = STRING: Unknown
.1.3.6.1.2.1.1.8.0 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.2.1 = OID: .1.3.6.1.6.3.11.3.1.1
.1.3.6.1.2.1.1.9.1.2.2 = OID: .1.3.6.1.6.3.15.2.1.1
.1.3.6.1.2.1.1.9.1.2.3 = OID: .1.3.6.1.6.3.10.3.1.1
.1.3.6.1.2.1.1.9.1.2.4 = OID: .1.3.6.1.6.3.1
.1.3.6.1.2.1.1.9.1.2.5 = OID: .1.3.6.1.6.3.16.2.2.1
.1.3.6.1.2.1.1.9.1.2.6 = OID: .1.3.6.1.2.1.49
.1.3.6.1.2.1.1.9.1.2.7 = OID: .1.3.6.1.2.1.4
.1.3.6.1.2.1.1.9.1.2.8 = OID: .1.3.6.1.2.1.50
.1.3.6.1.2.1.1.9.1.2.9 = OID: .1.3.6.1.6.3.13.3.1.3
.1.3.6.1.2.1.1.9.1.2.10 = OID: .1.3.6.1.2.1.92
.1.3.6.1.2.1.1.9.1.3.1 = STRING: The MIB for Message Processing and Dispatching.
.1.3.6.1.2.1.1.9.1.3.2 = STRING: The management information definitions for the SNMP User-based Security Model.
.1.3.6.1.2.1.1.9.1.3.3 = STRING: The SNMP Management Architecture MIB.
.1.3.6.1.2.1.1.9.1.3.4 = STRING: The MIB module for SNMPv2 entities
.1.3.6.1.2.1.1.9.1.3.5 = STRING: View-based Access Control Model for SNMP.
.1.3.6.1.2.1.1.9.1.3.6 = STRING: The MIB module for managing TCP implementations
.1.3.6.1.2.1.1.9.1.3.7 = STRING: The MIB module for managing IP and ICMP implementations
.1.3.6.1.2.1.1.9.1.3.8 = STRING: The MIB module for managing UDP implementations
.1.3.6.1.2.1.1.9.1.3.9 = STRING: The MIB modules for managing SNMP Notification, plus filtering.
.1.3.6.1.2.1.1.9.1.3.10 = STRING: The MIB module for logging SNMP Notifications.
.1.3.6.1.2.1.1.9.1.4.1 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.2 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.3 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.4 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.5 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.6 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.7 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.8 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.9 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.1.9.1.4.10 = Timeticks: (1) 0:00:00.01
.1.3.6.1.2.1.25.1.1.0 = Timeticks: (68440433) 7 days, 22:06:44.33
.1.3.6.1.2.1.25.1.2.0 = STRING: 2022-3-1,7:14:33.0,+13:0
[librenms@librenms ~]$snmpwalk -v2c -On -c $COMMUNITY $HOST .1.3.6.1.4.1.22408.1.1.2.1
.1.3.6.1.4.1.22408.1.1.2.1.2.118.109.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.3.99.112.117.1 = STRING: "29"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.49.1 = STRING: "1100"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.50.1 = STRING: "2300"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.51.1 = STRING: "2200"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.52.1 = STRING: "2200"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.53.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.4.102.97.110.54.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.4.118.100.98.49.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.4.118.100.98.50.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.49.1 = STRING: "0.85 0.53 0.48"
.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.50.1 = STRING: "0.85"
.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.51.1 = STRING: "0.53"
.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.52.1 = STRING: "0.48"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.49.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.50.1 = STRING: "clean"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.51.1 = STRING: "Total Devices : 2"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.52.1 = STRING: "2"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.53.1 = STRING: "Raid Devices : 2"
.1.3.6.1.4.1.22408.1.1.2.1.5.114.97.105.100.54.1 = STRING: "2"
.1.3.6.1.4.1.22408.1.1.2.1.6.109.97.105.110.116.49.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.6.109.97.105.110.116.50.1 = STRING: "Operational"
.1.3.6.1.4.1.22408.1.1.2.1.7.118.101.114.115.105.111.110.1 = STRING: "PrimeKeyAppliance.3.4.3"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.49.1 = STRING: "1"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.50.1 = STRING: "1"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.51.1 = STRING: "1"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.52.1 = STRING: "4"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.53.1 = STRING: "Synced"
.1.3.6.1.4.1.22408.1.1.2.1.8.99.108.117.115.116.101.114.54.1 = STRING: "3064"
.1.3.6.1.4.1.22408.1.1.2.1.8.104.101.97.108.116.104.101.49.1 = STRING: "ALLOK"
.1.3.6.1.4.1.22408.1.1.2.1.8.104.101.97.108.116.104.101.50.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.1.8.104.101.97.108.116.104.115.49.1 = STRING: "not-running"
.1.3.6.1.4.1.22408.1.1.2.1.8.104.101.97.108.116.104.115.50.1 = STRING: "1"
.1.3.6.1.4.1.22408.1.1.2.1.8.104.101.97.108.116.104.115.50.1 = No more variables left in this MIB View (It is past the end of the MIB tree)
[librenms@librenms ~]$snmpwalk -v2c -On -c $COMMUNITY $HOST .1.3.6.1.4.1.22408.1.1.2.2
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.49.1 = STRING: "STATUS_is_OPER"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.50.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.51.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.52.1 = STRING: "0.104 V"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.53.1 = STRING: "1"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.54.1 = STRING: "CS700862"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.55.1 = STRING: "External Battery: absence"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.56.1 = STRING: "0"
.1.3.6.1.4.1.22408.1.1.2.2.4.104.115.109.56.1 = No more variables left in this MIB View (It is past the end of the MIB tree)
[librenms@librenms ~]$
```

**Comment by bl3nd3r** [comment]
I’ve tried and failed to use Yaml for this. As you mentioned they’re not really tables we’re dealing with here, it’s perhaps a bit easier to see that via the snmp walk here: https://p.libren.ms/view/94902f64

In my view there’s not (yet) enough flexibility available via Yaml to deal with MiBs that are structured like this. 

I don’t work for PrimeKey, just one of their customers, so I can only work with the MiBs they’ve make available.

Is there anything I can do to help get this PR merged?

**Comment by bl3nd3r** [comment]
@PipoCanaja is there anything I can do to help speed up this review?

File: misc/discovery_schema.json [hunk 2]
```diff
@@ -393,7 +393,7 @@
                             },
                             "num_oid": {
                                 "type": "string",
-                                "pattern": "^(\\.\\d+)+\\.?\\{\\{ \\$index \\}\\}(\\.\\d+)*$"
```
Comment (bl3nd3r) [inline misc/discovery_schema.json:396]: This seems to be a necessary step in order to poll sensors using YAML if they don't come in table format.

**Comment by bl3nd3r** [comment]
> Hi @bl3nd3r
> Could you try to use as much YAML file as possible instead of PHP files ? When possible, YAML files are easier to maintain in the longterm than PHP files.
> Thanx

Hi @PipoCanaja & @kkrumm1 
Well I think I've done it. All the new PHP files have been converted to YAML.
I had to modify one test case slightly, and also deal with the fact that it's using snmpwalk to pull single OIDS - which works fine until you get to the end of the table.
Can you please review and ideally approve this PR?
Thanks

File: includes/common.php [hunk 3]
```diff
@@ -986,3 +986,17 @@ function array_by_column($array, $column)
...
+/**
+ * Return the first word from a string.
+ *
+ * For example if passed an SNMP response that contains units such as
+ * "3.115 V", this function will simply return "3.115".
+ *
+ * @param  string  $value
+ * @return string (containing only the first word of $value)
+ */
+function first_word($value)
```
Comment (murrant) [inline includes/common.php:999]: \LibreNMS\Util\Number::cast()

Does what you want, can you please us that?

Comment (murrant) [reply includes/common.php:999]: Actually, the sensor code should already use Number::cast().  Are you sure this is needed?

File: includes/definitions/discovery/primekey.yaml [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        count:
+            options:
+                skip_value_lt: 0
+            data:
+                -
+                    oid: PRIMEKEY-APPLIANCE-MIB::pkASfpLoad1m
+                    num_oid: '.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.50.1'
+                    descr: ' 1m CPU load'
+                    group: 'CPU'
+                    index: 1
```
Comment (murrant) [inline includes/definitions/discovery/primekey.yaml:18]: Hardcoding index like this only allows one sensor for each oid, but perhaps that is intended as I see you hard coded num_oid above too? 
Additionally, index can be a name.  The index field is kind of mislabled.  It should be called something like tag or uid.

File: misc/discovery_schema.json [hunk 5]
```diff
@@ -393,7 +393,7 @@
...
+                                "pattern": "^(\\.\\d+)+\\.?(\\{\\{ \\$index \\}\\})*(\\.\\d+)*$"
```
Comment (murrant) [inline misc/discovery_schema.json:396]: So, I checked and most or all? of these are returning an oid without an index.

This means instead of modifying the detection, you can just add {{ $index }} at the end of your oid and it will insert an empty string. Like so `num_oid: '.1.3.6.1.4.1.22408.1.1.2.1.5.108.111.97.100.50.1{{ $index }}'`

I know it feels a little weird, but this detection is here for good reason.

File: includes/common.php (hunk 3 above)
Comment (bl3nd3r) [reply includes/common.php:999]: The MiB returns a string that ends with the units (in Volts) with a space between the number and the units.
I raised a case with PrimeKey and they declined to change this, as it's related to how they get the voltage from the HSM.

In any case let's see what happens if I don't call that function.

Comment (murrant) [reply includes/common.php:999]:
```
$ php artisan tinker
>>> Number::cast('1.323 V')
=> 1.323
```

File: includes/definitions/discovery/primekey.yaml (hunk 4 above)
Comment (bl3nd3r) [reply includes/definitions/discovery/primekey.yaml:18]: So oid and num_oid can potentially have the same value? Let me respond to the next question before you answer this.

File: misc/discovery_schema.json (hunk 5 above)
Comment (bl3nd3r) [reply misc/discovery_schema.json:396]: All of these are returning an OID without an index. Therefore I suspect OID should remain as a name, and num_oid should be either the same name OR numeric description of the same OID with {{ $index}} tagged on the end.

File: includes/common.php (hunk 3 above)
Comment (bl3nd3r) [reply includes/common.php]: Number::cast works

File: includes/definitions/discovery/primekey.yaml (hunk 4 above)
Comment (bl3nd3r) [reply includes/definitions/discovery/primekey.yaml:18]: setting num_oid to numeric index followed by {{.$index }} works and avoids the need to change test cases

File: misc/discovery_schema.json (hunk 5 above)
Comment (bl3nd3r) [reply misc/discovery_schema.json]: setting num_oid to numeric index followed by {{.$index }} works and avoids the need to change test cases

File: includes/definitions/discovery/primekey.yaml (hunk 4 above)
Comment (bl3nd3r) [reply includes/definitions/discovery/primekey.yaml:18]: There is only ever one sensor for each oid, it seems PrimeKey don't use mib tables. 

I've tried to set the indices so that things order nicely, e.g. based on the order of the oids in the mib.

**Review by murrant — APPROVED** [review]

File: includes/definitions/discovery/primekey.yaml (hunk 4 above)
Comment (bl3nd3r) [reply includes/definitions/discovery/primekey.yaml:18]: Just found this:
https://github.com/librenms/librenms/pull/13861
Should I raise another PR to try and align my static indexes?
